src/circuit/gadget/utilities/lookup_range_check.rs

```diff
@@ -135,7 +131,14 @@ impl<F: FieldExt + PrimeFieldBits, const K: usize> LookupRangeCheckConfig<F, K>
                 // Copy `element` and initialize running sum `z_0 = element` to decompose it.
                 let z_0 = copy(&mut region, || "z_0", self.running_sum, 0, &element)?;
 
-                self.range_check(&mut region, z_0, num_words, strict)
+                let zs = self.range_check(&mut region, z_0, num_words)?;
+
+                if strict {
+                    // Constrain the final `z` to be zero.
+                    region.constrain_constant(zs.last().unwrap().cell(), F::zero())?;
+                }
+
+                Ok(zs)
```

Comment:
This can be un-refactored now we're using `Region::constrain_constant` instead of public inputs:
```suggestion
                self.range_check(&mut region, z_0, num_words, strict)
```
(These changes here and below should be squashed into the `constants` commit, after review is complete and ACKed, so we can confirm the squash is a no-op.)